File: leetcode/0001-1000/0013-roman-to-integer/RomanToInteger.ts
function romanToInt(s: string): number {
  const symbols = {
    "M": 1000,
    "D": 500,
    "C": 100,
    "L": 50,
    "X": 10,
    "V": 5,
    "I": 1,
  };

  let value = 0;

  for (let i = 0; i < s.length; i++) {
    if (i < s.length - 1 && symbols[s[i]] < symbols[s[i + 1]]) {
      value += symbols[s[i + 1]] - symbols[s[i]];
      i++;
    } else {
      value += symbols[s[i]];
    }
  }

  return value;
};

console.log(romanToInt("III"));
console.log(romanToInt("XXIX"));

File: leetcode/0001-1000/0001-0100/0013-roman-to-integer/RomanToInteger.java
public class RomanToInteger {
  public static void main(String[] args) {
    System.out.println(new Solution().romanToInt("III"));
  }
}

class Solution {
  public int romanToInt(String s) {
    int[] values = { 1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1 };
    String[] symbols = { "M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I" };

    int i = 0;
    int v = 0;

    for (int j = 0; j < symbols.length; j++) {
      while (i < s.length() && i + symbols[j].length() <= s.length()
          && symbols[j].equals(s.substring(i, i + symbols[j].length()))) {
        v += values[j];
        i += symbols[j].length();
      }
    }

    return v;
  }
}
